src/bandwidth_limit.cpp

```diff
@@ -59,18 +59,27 @@ namespace libtorrent {
 
 	void bandwidth_channel::update_quota(int dt_milliseconds)
 	{
+		TORRENT_ASSERT_VAL(m_limit >= 0, m_limit);
+
+		TORRENT_ASSERT_VAL(m_limit < inf, m_limit);
+
 		if (m_limit == 0) return;
 
-		// avoid integer overflow
-		if (m_limit >= std::numeric_limits<int>::max() / dt_milliseconds)
+		// "to_add" should never have int64 overflow: "m_limit" contains < "<int>::max"
+		std::int64_t to_add = (m_limit * dt_milliseconds + 500) / 1000;
```

Comment:
I don't think it's necessary to widen to 64 bits here. you seem to prefer that, is there a compelling reason?
also, the multiplication is still performed in 32 bits, you need to widen one of the operands before multiplying if you want it to be 64 bits.